tools/clang/lib/SPIRV/GlPerVertex.cpp

```diff
@@ -0,0 +1,751 @@
+//===--- GlPerVertex.cpp - GlPerVertex implementation ------------*- C++ -*-==//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+#include "GlPerVertex.h"
+
+#include "clang/AST/HlslTypes.h"
+
+namespace clang {
+namespace spirv {
+
+namespace {
+/// \brief Returns true if the given decl has a semantic string attached and
+/// writes the info to *semanticStr, *semantic, and *semanticIndex.
+// TODO: duplication! Same as the one in DeclResultIdMapper.cpp
+bool getStageVarSemantic(const NamedDecl *decl, llvm::StringRef *semanticStr,
+                         const hlsl::Semantic **semantic,
+                         uint32_t *semanticIndex) {
+  for (auto *annotation : decl->getUnusualAnnotations()) {
+    if (auto *sema = dyn_cast<hlsl::SemanticDecl>(annotation)) {
+      *semanticStr = sema->SemanticName;
+      llvm::StringRef semanticName;
+      hlsl::Semantic::DecomposeNameAndIndex(*semanticStr, &semanticName,
+                                            semanticIndex);
+      *semantic = hlsl::Semantic::GetByName(semanticName);
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Returns the type of the given decl. If the given decl is a FunctionDecl,
+/// returns its result type.
+inline QualType getTypeOrFnRetType(const DeclaratorDecl *decl) {
```

Comment:
also a duplicate function

Reply:
This is inlined function and quite short. So doesn't matter that much IMHO.